File: python/src/acedatacloud/resources/oauth.py
"""OAuth 2.0 client helpers for "Sign in with Ace Data Cloud"."""

from __future__ import annotations

from typing import Any

import httpx

_AUTH_BASE = "https://auth.acedata.cloud"


class OAuth:
    """Synchronous OAuth 2.0 helper for "Sign in with Ace Data Cloud"."""

    def __init__(self, auth_base_url: str = _AUTH_BASE) -> None:
        self._auth_base_url = auth_base_url.rstrip("/")
        self._client = httpx.Client()

    def build_authorization_url(
        self,
        *,
        client_id: str,
        redirect_uri: str,
        scope: str,
        state: str,
        code_challenge: str | None = None,
        code_challenge_method: str = "S256",
        response_type: str = "code",
    ) -> str:
        params: dict[str, str] = {
            "response_type": response_type,
            "client_id": client_id,
            "redirect_uri": redirect_uri,
            "scope": scope,
            "state": state,
        }
        if code_challenge is not None:
            params["code_challenge"] = code_challenge
            params["code_challenge_method"] = code_challenge_method
        query = httpx.URL("", params=params).query.decode()
        return f"{self._auth_base_url}/oauth2/authorize?{query}"

    def discover(self) -> dict[str, Any]:
        resp = self._client.get(f"{self._auth_base_url}/.well-known/oauth-authorization-server")
        resp.raise_for_status()
        return resp.json()

    def exchange_code(
        self,
        *,
        client_id: str,
        code: str,
        redirect_uri: str,
        client_secret: str | None = None,
        code_verifier: str | None = None,
    ) -> dict[str, Any]:
        data: dict[str, str] = {
            "grant_type": "authorization_code",
            "code": code,
            "client_id": client_id,
            "redirect_uri": redirect_uri,
        }
        if client_secret is not None:
            data["client_secret"] = client_secret
        if code_verifier is not None:
            data["code_verifier"] = code_verifier
        resp = self._client.post(f"{self._auth_base_url}/oauth2/token", data=data)
        resp.raise_for_status()
        return resp.json()

    def refresh_token(
        self,
        *,
        refresh_token: str,
        client_id: str | None = None,
        client_secret: str | None = None,
    ) -> dict[str, Any]:
        data: dict[str, str] = {
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
        }
        if client_id is not None:
            data["client_id"] = client_id
        if client_secret is not None:
            data["client_secret"] = client_secret
        resp = self._client.post(f"{self._auth_base_url}/oauth2/token", data=data)
        resp.raise_for_status()
        return resp.json()

    def revoke_token(
        self,
        *,
        token: str,
        client_id: str | None = None,
        client_secret: str | None = None,
    ) -> None:
        data: dict[str, str] = {"token": token}
        if client_id is not None:
            data["client_id"] = client_id
        if client_secret is not None:
            data["client_secret"] = client_secret
        self._client.post(f"{self._auth_base_url}/oauth2/revoke", data=data)

    def get_user_info(self, access_token: str) -> dict[str, Any]:
        resp = self._client.get(
            f"{self._auth_base_url}/api/v1/users/me",
            headers={"authorization": "Bearer " + access_token},
        )
        resp.raise_for_status()
        return resp.json()

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> OAuth:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()


class AsyncOAuth:
    """Async OAuth 2.0 helper for "Sign in with Ace Data Cloud"."""

    def __init__(self, auth_base_url: str = _AUTH_BASE) -> None:
        self._auth_base_url = auth_base_url.rstrip("/")
        self._client = httpx.AsyncClient()

    def build_authorization_url(
        self,
        *,
        client_id: str,
        redirect_uri: str,
        scope: str,
        state: str,
        code_challenge: str | None = None,
        code_challenge_method: str = "S256",
        response_type: str = "code",
    ) -> str:
        params: dict[str, str] = {
            "response_type": response_type,
            "client_id": client_id,
            "redirect_uri": redirect_uri,
            "scope": scope,
            "state": state,
        }
        if code_challenge is not None:
            params["code_challenge"] = code_challenge
            params["code_challenge_method"] = code_challenge_method
        query = httpx.URL("", params=params).query.decode()
        return f"{self._auth_base_url}/oauth2/authorize?{query}"

    async def discover(self) -> dict[str, Any]:
        resp = await self._client.get(f"{self._auth_base_url}/.well-known/oauth-authorization-server")
        resp.raise_for_status()
        return resp.json()

    async def exchange_code(
        self,
        *,
        client_id: str,
        code: str,
        redirect_uri: str,
        client_secret: str | None = None,
        code_verifier: str | None = None,
    ) -> dict[str, Any]:
        data: dict[str, str] = {
            "grant_type": "authorization_code",
            "code": code,
            "client_id": client_id,
            "redirect_uri": redirect_uri,
        }
        if client_secret is not None:
            data["client_secret"] = client_secret
        if code_verifier is not None:
            data["code_verifier"] = code_verifier
        resp = await self._client.post(f"{self._auth_base_url}/oauth2/token", data=data)
        resp.raise_for_status()
        return resp.json()

    async def refresh_token(
        self,
        *,
        refresh_token: str,
        client_id: str | None = None,
        client_secret: str | None = None,
    ) -> dict[str, Any]:
        data: dict[str, str] = {
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
        }
        if client_id is not None:
            data["client_id"] = client_id
        if client_secret is not None:
            data["client_secret"] = client_secret
        resp = await self._client.post(f"{self._auth_base_url}/oauth2/token", data=data)
        resp.raise_for_status()
        return resp.json()

    async def revoke_token(
        self,
        *,
        token: str,
        client_id: str | None = None,
        client_secret: str | None = None,
    ) -> None:
        data: dict[str, str] = {"token": token}
        if client_id is not None:
            data["client_id"] = client_id
        if client_secret is not None:
            data["client_secret"] = client_secret
        await self._client.post(f"{self._auth_base_url}/oauth2/revoke", data=data)

    async def get_user_info(self, access_token: str) -> dict[str, Any]:
        resp = await self._client.get(
            f"{self._auth_base_url}/api/v1/users/me",
            headers={"authorization": "Bearer " + access_token},
        )
        resp.raise_for_status()
        return resp.json()

    async def close(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> AsyncOAuth:
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()

File: typescript/src/resources/oauth.ts
/** OAuth 2.0 client helpers for "Sign in with Ace Data Cloud". */

const AUTH_BASE = 'https://auth.acedata.cloud';

export interface OAuthTokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
  scope: string;
  refresh_token?: string;
}

export interface OAuthUserInfo {
  id: string;
  username?: string;
  nickname?: string;
  avatar?: string;
  is_verified?: boolean;
  date_joined?: string;
  email?: string;
  phone?: string;
  region?: string;
}

export interface BuildAuthorizationUrlParams {
  clientId: string;
  redirectUri: string;
  scope: string;
  state: string;
  codeChallenge?: string;
  codeChallengeMethod?: 'S256' | 'plain';
  responseType?: string;
}

export interface ExchangeCodeParams {
  clientId: string;
  code: string;
  redirectUri: string;
  clientSecret?: string;
  codeVerifier?: string;
}

export interface RefreshTokenParams {
  refreshToken: string;
  clientId?: string;
  clientSecret?: string;
}

export interface RevokeTokenParams {
  token: string;
  clientId?: string;
  clientSecret?: string;
}

async function checkResponse(resp: Response): Promise<void> {
  if (!resp.ok) {
    const text = await resp.text();
    let message = text;
    try {
      const body = JSON.parse(text) as Record<string, unknown>;
      if (typeof body.error_description === 'string') message = body.error_description;
      else if (typeof body.error === 'string') message = body.error;
    } catch {
      // use raw text as message
    }
    throw new Error(`OAuth request failed (${resp.status}): ${message}`);
  }
}

export class OAuth {
  private authBaseURL: string;

  constructor(authBaseURL?: string) {
    let base = authBaseURL ?? AUTH_BASE;
    while (base.endsWith('/')) base = base.slice(0, -1);
    this.authBaseURL = base;
  }

  buildAuthorizationUrl(params: BuildAuthorizationUrlParams): string {
    const urlParams = new URLSearchParams({
      response_type: params.responseType ?? 'code',
      client_id: params.clientId,
      redirect_uri: params.redirectUri,
      scope: params.scope,
      state: params.state,
    });
    if (params.codeChallenge) {
      urlParams.set('code_challenge', params.codeChallenge);
      urlParams.set('code_challenge_method', params.codeChallengeMethod ?? 'S256');
    }
    return `${this.authBaseURL}/oauth2/authorize?${urlParams.toString()}`;
  }

  async discover(): Promise<Record<string, unknown>> {
    const resp = await fetch(`${this.authBaseURL}/.well-known/oauth-authorization-server`);
    await checkResponse(resp);
    return (await resp.json()) as Record<string, unknown>;
  }

  async exchangeCode(params: ExchangeCodeParams): Promise<OAuthTokenResponse> {
    const body = new URLSearchParams({
      grant_type: 'authorization_code',
      code: params.code,
      client_id: params.clientId,
      redirect_uri: params.redirectUri,
    });
    if (params.clientSecret) body.set('client_secret', params.clientSecret);
    if (params.codeVerifier) body.set('code_verifier', params.codeVerifier);

    const resp = await fetch(`${this.authBaseURL}/oauth2/token`, {
      method: 'POST',
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    });
    await checkResponse(resp);
    return (await resp.json()) as OAuthTokenResponse;
  }

  async refreshToken(params: RefreshTokenParams): Promise<OAuthTokenResponse> {
    const body = new URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: params.refreshToken,
    });
    if (params.clientId) body.set('client_id', params.clientId);
    if (params.clientSecret) body.set('client_secret', params.clientSecret);

    const resp = await fetch(`${this.authBaseURL}/oauth2/token`, {
      method: 'POST',
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    });
    await checkResponse(resp);
    return (await resp.json()) as OAuthTokenResponse;
  }

  async revokeToken(params: RevokeTokenParams): Promise<void> {
    const body = new URLSearchParams({ token: params.token });
    if (params.clientId) body.set('client_id', params.clientId);
    if (params.clientSecret) body.set('client_secret', params.clientSecret);

    const resp = await fetch(`${this.authBaseURL}/oauth2/revoke`, {
      method: 'POST',
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    });
    await checkResponse(resp);
  }

  async getUserInfo(accessToken: string): Promise<OAuthUserInfo> {
    const resp = await fetch(`${this.authBaseURL}/api/v1/users/me`, {
      headers: { authorization: 'Bearer ' + accessToken },
    });
    await checkResponse(resp);
    return (await resp.json()) as OAuthUserInfo;
  }
}
